### DISCUSSION TITLE
Question in regards to potential version conflct and support TF versions

### DISCUSSION THREAD

**Discussion by thungp** [Trusted-AI/adversarial-robustness-toolbox#1286] [Q&A] [answered]
1. If I'm reading the requirements.txt (https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/00206ee5678631ac05e430c04b46c2163d06dcf2/requirements.txt

scipy==1.7.1
 # supported versions: (tensorflow==2.2.0 with keras==2.3.1) or (tensorflow==1.15.4 with keras==2.2.5)) file correctly, 
tensorflow vers 2.2.0 and scipy==1.7.1 should be able to coexist.
I tried to pin 
tensorflow==2.2.0
and then do a 
pip install -r <ARTS requirements file>
and pip reported back with the following:
ERROR: Cannot install -r requirements_tf_2.2.0_v1.txt (line 11), -r requirements_tf_2.2.0_v1.txt (line 15), -r requirements_tf_2.2.0_v1.txt (line 23), -r requirements_tf_2.2.0_v1.txt (line 9) and scipy==1.7.1 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested scipy==1.7.1
    statsmodels 0.12.2 depends on scipy>=1.1
    resampy 0.2.2 depends on scipy>=0.13
    librosa 0.8.1 depends on scipy>=1.0.0
    tensorflow 2.2.0 depends on scipy==1.4.1; python_version >= "3"

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: Resolution Impossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
Note: 
WARNING: You are using pip version 21.2.3;

I am assuming that something is requiring scipy==1.7.1
(Which (FYSA) least according to the scipy release notes (https://github.com/scipy/scipy/releases)
states we need to be using This release requires Python 3.7+ ...)

Out of ignorance, is there a way to see what particular package is actually calling for scipy==1.7.1?

Curious if anyone else is seeing the error?

Out of curiosity I tried to upgrade to tensorflow==2.3.0  # since that is what I am currently using
but pip is also giving me the error:
ERROR: Cannot install -r requirements_tf_2.3.0_v1.txt (line 11), -r requirements_tf_2.3.0_v1.txt (line 15), -r requirements_tf_2.3.0_v1.txt (line 23), -r requirements_tf_2.3.0_v1.txt (line 9) and scipy==1.7.1 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested scipy==1.7.1
    statsmodels 0.12.2 depends on scipy>=1.1
    resampy 0.2.2 depends on scipy>=0.13
    librosa 0.8.1 depends on scipy>=1.0.0
    tensorflow 2.3.0 depends on scipy==1.4.1

I was curious about the tensorflows depenency on scipy and came across this link
https://github.com/streamlit/streamlit/pull/2193
Which if I'm understanding it correctly, it may not be a real dependency?

I then went and tried to pin 
scipy==1.4.1
with tensorflow 2.3.0 to see if that my find some sort of resolution to create a working environment. 
an it appears to at least create the environment. 

As an asside, the notebook example, I'm actually interested in getting working ideally with tf 2.3
is https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/adversarial_training_mnist.ipynb

**Comment by beat-buesser** [comment, marked as answer]
Hi @thungp Thank you very much for your interest in ART! Currently the `requirements.txt` file is only used to define the base versions for our CI and it is not intended to be used to install all dependencies for ART because each CI run makes selected changes to it (resolving the issues you have described above). The dependencies of ART have very large ranges and the `setup.py` provides install options to automate the installation of dependencies with more information in our Wiki: https://github.com/Trusted-AI/adversarial-robustness-toolbox/wiki/Get-Started
